### PR TITLE
sys: onoff: support clients using synchronous transitions

### PR DESCRIPTION
**Pull request by pabigot**
*The use case for this is in #27360 where it allows simplified handling of GPIO-controlled regulators when an on-SoC GPIO is used with no start/stop delays.*

The on-off manager infrastructure is designed to support robust asynchronous transition between binary states where multiple clients may be initiating a transition from any context.  The actual transition is performed using a manager that tracks the current state and pending operations.  Requests are initiated by passing a reference to an onoff_client object that holds client state including the notification mechanism.

This API may be used in subsystems where the transitions for a particular driver are always synchronous and isr-ok, e.g. setting a SoC-controlled GPIO.  In this situation the full on-off manager infrastructure is wasteful.  All we need is a record of the service state: off, active count, or error.

Add a data structure and an API that can be used to replace the onoff manager functionality in a situation where all transitions are isr-ok and synchronous while retaining compatible behavior from the client perspective.

### CODE REVIEW

File: lib/os/onoff.c [hunk 1]
```diff
@@ -600,3 +600,51 @@ int onoff_monitor_unregister(struct onoff_manager *mgr,
...
+	}
+	if (res < 0) {
+		srv->count = res;
+		state = ONOFF_STATE_ERROR;
+	} else if (on) {
+		srv->count += 1;
+		state = ONOFF_STATE_ON;
+	} else {
+		srv->count -= 1;
+		state = ONOFF_STATE_OFF;
```
Comment (nordic-krch) [inline lib/os/onoff.c:634]: should it go to `OFF` state only when count reaches 0?

File: include/sys/onoff.h [hunk 2]
```diff
@@ -560,6 +561,73 @@ int onoff_monitor_register(struct onoff_manager *mgr,
...
+ * notification using @p res.
+ *
+ * If the service was in an error state when locked, and @p res is non-negative
+ * when finalized, the count is reset to zero before completing finalization.
+ *
+ * @param srv pointer to the synchronous service state
+ *
+ * @param key the key returned by the preceding invocation of onoff_sync_lock().
+ *
+ * @param cli pointer to the onoff client through which completion information
```
Comment (nordic-krch) [inline include/sys/onoff.h:611]: it's not mentioned that cli must be null when `on` is false.

File: include/sys/onoff.h [hunk 3]
```diff
@@ -560,6 +561,73 @@ int onoff_monitor_register(struct onoff_manager *mgr,
...
+ * @param keyp pointer to where the lock key should be stored
+ *
+ * @return negative if the service is in an error state, otherwise the number of
+ * active requests at the time the lock was taken.
+ */
+int onoff_sync_lock(struct onoff_sync_service *srv,
+		    k_spinlock_key_t *keyp);
+
+/**
+ * @brief Process the completion of a transition in a synchronous service.
```
Comment (nordic-krch) [inline include/sys/onoff.h:598]: `... and unlocks the context` or something similar.

Comment (pabigot) [reply include/sys/onoff.h:598]: Part of processing is releasing the lock; it's mentioned in the first sentence after the brief.  I'll rework to address the other comments.

File: lib/os/onoff.c (hunk 1 above)
Comment (pabigot) [reply lib/os/onoff.c:634]: Yes.  However if control flow reaches this point `state` isn't used in what follows.  I'll replace this with a comment.

File: include/sys/onoff.h (hunk 3 above)
Comment (nordic-krch) [reply include/sys/onoff.h:600]: my bad, in fact it is mentioned. However, `onoff_sync_lock` mentions that in the brief section.

**Comment by pabigot** [comment]
@nordic-krch comments addressed and a note added to documentation about the role of this extension.

**Review by nordic-krch — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**